File: evaluations/pypsa/tests/extensibility/test_b4_stochastic_wrapping_small.py
"""
Test B-4: Stochastic Wrapping -- Generate 20 scenarios with correlated load/renewable
perturbations by resource type, solve 12hr multi-period DCOPF per scenario, collect
prices and dispatch.

Dimension: extensibility
Network: SMALL (ACTIVSg 2000-bus)
Pass condition: Tool accepts timeseries inputs programmatically. Scenario loop
    expressible without excessive per-scenario overhead. Results collectable in
    structured format.
Tool: PyPSA 1.1.2
"""

from __future__ import annotations

import time
import traceback
from pathlib import Path

import numpy as np
import pandas as pd

REPO_ROOT = Path(__file__).resolve().parent.parent.parent.parent.parent
NETWORK_FILE = str(REPO_ROOT / "data" / "networks" / "case_ACTIVSg2000.m")

SOLVER_NAME = "highs"
SOLVER_OPTIONS = {
    "time_limit": 300,
    "presolve": "on",
    "threads": 1,
    "output_flag": False,
}

N_SCENARIOS = 20
N_HOURS = 12
RNG_SEED = 42


def _load_network_with_costs(case_path: str):
    """Load MATPOWER .m file and set marginal costs from gencost."""
    import pypsa
    from matpowercaseframes import CaseFrames

    cf = CaseFrames(case_path)
    ppc = {
        "version": "2",
        "baseMVA": cf.baseMVA,
        "bus": cf.bus.values,
        "gen": cf.gen.values,
        "branch": cf.branch.values,
    }

    net = pypsa.Network()
    net.import_from_pypower_ppc(ppc)

    gencost = cf.gencost.values
    rng = np.random.default_rng(seed=99)
    for i, gen_idx in enumerate(net.generators.index):
        if i < len(gencost):
            cost_row = gencost[i]
            cost_type = int(cost_row[0])
            n_coeffs = int(cost_row[3])
            if cost_type == 2 and n_coeffs >= 2:
                c1 = float(cost_row[4 + n_coeffs - 2])
                perturbation = rng.uniform(-0.1, 0.1) * c1
                net.generators.loc[gen_idx, "marginal_cost"] = c1 + perturbation

    return net


def _classify_generators(n):
    """Classify generators by cost quartile as resource types."""
    costs = n.generators["marginal_cost"].sort_values()
    q25 = costs.quantile(0.25)
    q75 = costs.quantile(0.75)
    types = {}
    for gen in n.generators.index:
        mc = n.generators.loc[gen, "marginal_cost"]
        if mc <= q25:
            types[gen] = "baseload"
        elif mc >= q75:
            types[gen] = "peaker"
        else:
            types[gen] = "intermediate"
    return types


def _generate_scenarios(rng, wind_base, solar_base, n_scenarios):
    """Generate correlated perturbations, independent by resource type."""
    scenarios = []
    for s in range(n_scenarios):
        load_factor = rng.uniform(0.92, 1.08)
        baseload_factor = rng.uniform(0.97, 1.03)
        intermediate_factor = rng.uniform(0.93, 1.07)
        peaker_factor = rng.uniform(0.90, 1.10)
        wind_factor = rng.uniform(0.70, 1.30)
        solar_factor = rng.uniform(0.80, 1.20)

        scenarios.append(
            {
                "load_factor": load_factor,
                "gen_factors": {
                    "baseload": baseload_factor,
                    "intermediate": intermediate_factor,
                    "peaker": peaker_factor,
                },
                "wind_profile": np.clip(wind_base * wind_factor, 0, 1),
                "solar_profile": np.clip(solar_base * solar_factor, 0, 1),
            }
        )
    return scenarios


def run(network_file: str = NETWORK_FILE) -> dict:
    """Execute stochastic wrapping test on 2000-bus and return structured results."""
    results = {
        "status": "fail",
        "wall_clock_seconds": 0.0,
        "details": {},
        "errors": [],
        "workarounds": [],
    }

    start_total = time.perf_counter()
    try:
        # 1. Load base network
        n_base = _load_network_with_costs(network_file)

        # Set up 12 hourly snapshots
        snapshots = pd.date_range("2024-01-01", periods=N_HOURS, freq="h")
        n_base.set_snapshots(snapshots)

        # Set time-varying loads
        base_loads = n_base.loads["p_set"].copy()
        load_df = pd.DataFrame(
            {load_idx: base_loads[load_idx] for load_idx in n_base.loads.index},
            index=snapshots,
        )
        n_base.loads_t.p_set = load_df

        # Add renewables at buses with generators
        gen_buses = n_base.generators["bus"].unique()
        wind_bus = gen_buses[0] if len(gen_buses) > 0 else n_base.buses.index[0]
        solar_bus = gen_buses[min(5, len(gen_buses) - 1)] if len(gen_buses) > 5 else wind_bus

        n_base.add(
            "Generator",
            "Wind_1",
            bus=wind_bus,
            p_nom=500.0,
            marginal_cost=0.0,
            carrier="wind",
        )
        n_base.add(
            "Generator",
            "Solar_1",
            bus=solar_bus,
            p_nom=400.0,
            marginal_cost=0.0,
            carrier="solar",
        )

        hours = np.arange(N_HOURS)
        wind_base = 0.3 + 0.15 * np.sin(2 * np.pi * hours / 24)
        solar_base = np.clip(0.5 * np.sin(np.pi * (hours - 4) / 12), 0, 1)

        p_max_pu = pd.DataFrame(1.0, index=snapshots, columns=n_base.generators.index)
        p_max_pu["Wind_1"] = wind_base
        p_max_pu["Solar_1"] = solar_base
        n_base.generators_t.p_max_pu = p_max_pu

        # Classify generators
        gen_types = _classify_generators(n_base)

        # 2. Generate scenarios
        rng = np.random.default_rng(seed=RNG_SEED)
        scenarios = _generate_scenarios(rng, wind_base, solar_base, N_SCENARIOS)

        # 3. Solve each scenario
        all_objectives = []
        scenario_times = []
        all_lmp_stats = []

        results["workarounds"].append(
            "Manually set marginal_cost from gencost data (PPC importer does not import gencost)"
        )

        for s_idx, scenario in enumerate(scenarios):
            n = n_base.copy()

            # Apply load perturbation
            load_scenario = pd.DataFrame(
                {
                    load_idx: base_loads[load_idx] * scenario["load_factor"]
                    for load_idx in n.loads.index
                },
                index=snapshots,
            )
            n.loads_t.p_set = load_scenario

            # Apply generator p_nom perturbations
            for gen_idx in n.generators.index:
                if gen_idx in gen_types:
                    gtype = gen_types[gen_idx]
                    if gtype in scenario["gen_factors"]:
                        factor = scenario["gen_factors"][gtype]
                        p_nom = n_base.generators.loc[gen_idx, "p_nom"]
                        n.generators.loc[gen_idx, "p_nom"] = p_nom * factor

            # Apply renewable profiles
            p_max_pu_s = pd.DataFrame(1.0, index=snapshots, columns=n.generators.index)
            p_max_pu_s["Wind_1"] = scenario["wind_profile"]
            p_max_pu_s["Solar_1"] = scenario["solar_profile"]
            n.generators_t.p_max_pu = p_max_pu_s

            # Solve
            t0 = time.perf_counter()
            status = n.optimize(solver_name=SOLVER_NAME, solver_options=SOLVER_OPTIONS)
            t1 = time.perf_counter()
            scenario_times.append(t1 - t0)

            converged = "ok" in str(status).lower() or "optimal" in str(status).lower()
            if not converged:
                results["errors"].append(f"Scenario {s_idx} failed: {status}")
                continue

            all_objectives.append(float(n.objective))
            lmps = n.buses_t.marginal_price
            all_lmp_stats.append(
                {
                    "mean": float(lmps.values.mean()),
                    "min": float(lmps.values.min()),
                    "max": float(lmps.values.max()),
                }
            )

        elapsed_total = time.perf_counter() - start_total

        # 4. Aggregate results
        n_solved = len(all_objectives)
        if n_solved == 0:
            results["errors"].append("No scenarios converged")
            results["wall_clock_seconds"] = elapsed_total
            return results

        obj_arr = np.array(all_objectives)
        time_arr = np.array(scenario_times)

        results["details"] = {
            "n_scenarios_requested": N_SCENARIOS,
            "n_scenarios_solved": n_solved,
            "n_hours": N_HOURS,
            "solver": SOLVER_NAME,
            "total_time_s": elapsed_total,
            "solve_time_total_s": float(time_arr.sum()),
            "solve_time_mean_s": float(time_arr.mean()),
            "solve_time_std_s": float(time_arr.std()),
            "solve_time_min_s": float(time_arr.min()),
            "solve_time_max_s": float(time_arr.max()),
            "objective_mean": float(obj_arr.mean()),
            "objective_std": float(obj_arr.std()),
            "objective_min": float(obj_arr.min()),
            "objective_max": float(obj_arr.max()),
            "lmp_stats_sample": all_lmp_stats[:3],
            "api_mechanism": (
                "n.copy() for per-scenario network, "
                "direct DataFrame assignment for timeseries, "
                "n.optimize() per scenario"
            ),
            "network_stats": {
                "buses": len(n_base.buses),
                "generators": len(n_base.generators),
                "lines": len(n_base.lines),
            },
            "renewable_generators_added": ["Wind_1", "Solar_1"],
        }

        results["wall_clock_seconds"] = elapsed_total

        if n_solved == N_SCENARIOS:
            results["status"] = "pass"
        elif n_solved >= N_SCENARIOS * 0.9:
            results["status"] = "qualified_pass"

    except Exception as e:
        results["errors"].append(f"{type(e).__name__}: {e}")
        results["details"]["traceback"] = traceback.format_exc()
        results["wall_clock_seconds"] = time.perf_counter() - start_total

    return results


if __name__ == "__main__":
    import json

    result = run()
    print(json.dumps(result, indent=2, default=str))

File: evaluations/powermodels/tests/extensibility/test_b4_stochastic_wrapping_small.jl
#=
Test B-4: Stochastic Wrapping -- 20 scenarios of 12hr multi-period DCOPF
Dimension: extensibility
Network: SMALL (ACTIVSg 2000-bus)
Pass condition: Tool accepts timeseries inputs programmatically. Scenario loop
    expressible without excessive overhead. Results collectable.
Tool: PowerModels.jl v0.21.5
Solver: HiGHS
=#

using PowerModels, JuMP, HiGHS, JSON
using Random

function run(
    network_file::String=joinpath(
        @__DIR__, "..", "..", "..", "..", "data", "networks", "case_ACTIVSg2000.m"
    ),
)
    results = Dict(
        "status" => "fail",
        "wall_clock_seconds" => 0.0,
        "details" => Dict(),
        "errors" => String[],
        "workarounds" => String[],
    )

    t0 = time()
    try
        data = PowerModels.parse_file(network_file)

        # Fix generators with empty cost arrays
        for (id, gen) in data["gen"]
            if gen["model"] == 2 && (gen["ncost"] == 0 || isempty(gen["cost"]))
                gen["cost"] = [0.0, 0.0, 0.0]
                gen["ncost"] = 3
            end
        end

        results["details"]["num_buses"] = length(data["bus"])
        results["details"]["num_branches"] = length(data["branch"])
        results["details"]["num_generators"] = length(data["gen"])

        optimizer = JuMP.optimizer_with_attributes(
            HiGHS.Optimizer,
            "time_limit" => 300.0,
            "presolve" => "on",
            "threads" => 1,
            "output_flag" => false,
        )

        # ---- Scenario generation ----
        T = 12  # hours
        N_scenarios = 20
        rng = MersenneTwister(42)

        # Base hourly load profile (fraction of peak)
        base_profile = [0.80, 0.75, 0.72, 0.70, 0.75, 0.85, 0.95, 1.00, 0.98, 0.95, 0.90, 0.85]

        load_ids = collect(keys(data["load"]))
        gen_ids = sort(parse.(Int, collect(keys(data["gen"]))))

        # Classify generators: first 80% thermal, last 20% renewable-like
        n_thermal = Int(floor(0.8 * length(gen_ids)))
        thermal_gens = [string(g) for g in gen_ids[1:n_thermal]]
        renewable_gens = [string(g) for g in gen_ids[(n_thermal + 1):end]]

        scenarios = []
        for s in 1:N_scenarios
            load_common_factor = 1.0 + 0.03 * randn(rng)
            bus_noise = Dict(lid => 1.0 + 0.005 * randn(rng) for lid in load_ids)
            hourly_load = [base_profile[t] * load_common_factor for t in 1:T]

            thermal_avail = Dict(
                g => max(0.8, min(1.0, 1.0 + 0.03 * randn(rng))) for g in thermal_gens
            )
            renewable_avail = Dict(
                g => max(0.6, min(1.0, 0.9 + 0.08 * randn(rng))) for g in renewable_gens
            )

            push!(
                scenarios,
                Dict(
                    "load_common_factor" => load_common_factor,
                    "bus_noise" => bus_noise,
                    "hourly_load" => hourly_load,
                    "thermal_avail" => thermal_avail,
                    "renewable_avail" => renewable_avail,
                ),
            )
        end

        results["details"]["num_scenarios"] = N_scenarios
        results["details"]["num_periods"] = T
        results["details"]["scenario_generation_method"] = "Correlated perturbations by resource type"

        # ---- Warm-up solve ----
        try
            _d = deepcopy(data)
            _mn = PowerModels.replicate(_d, 2)
            PowerModels.solve_mn_opf(_mn, DCPPowerModel, optimizer)
        catch
            ;
        end

        # ---- Solve each scenario ----
        scenario_results = Dict{Int,Dict}()
        solve_times = Float64[]
        all_objectives = Float64[]

        t_loop = time()
        for s in 1:N_scenarios
            sc = scenarios[s]

            sc_data = deepcopy(data)

            for (g, avail) in sc["thermal_avail"]
                sc_data["gen"][g]["pmax"] *= avail
            end
            for (g, avail) in sc["renewable_avail"]
                sc_data["gen"][g]["pmax"] *= avail
            end

            mn_data = PowerModels.replicate(sc_data, T)

            for t in 1:T
                nw = mn_data["nw"][string(t)]
                for (lid, load) in nw["load"]
                    load["pd"] *= sc["hourly_load"][t] * get(sc["bus_noise"], lid, 1.0)
                    load["qd"] *= sc["hourly_load"][t] * get(sc["bus_noise"], lid, 1.0)
                end
            end

            t_solve = time()
            mn_result = PowerModels.solve_mn_opf(mn_data, DCPPowerModel, optimizer)
            dt_solve = time() - t_solve
            push!(solve_times, dt_solve)

            term = string(mn_result["termination_status"])

            if term in ["OPTIMAL", "LOCALLY_SOLVED", "ALMOST_LOCALLY_SOLVED"]
                push!(all_objectives, mn_result["objective"])
                scenario_results[s] = Dict(
                    "termination" => term,
                    "objective" => mn_result["objective"],
                    "solve_time" => dt_solve,
                )
            else
                scenario_results[s] = Dict(
                    "termination" => term, "objective" => nothing, "solve_time" => dt_solve
                )
            end

            println("Scenario $s: $term ($(round(dt_solve, digits=2))s)")
        end
        loop_time = time() - t_loop

        # ---- Collect aggregate results ----
        n_optimal = count(
            s -> scenario_results[s]["termination"] in ["OPTIMAL", "LOCALLY_SOLVED"], 1:N_scenarios
        )

        results["details"]["scenarios_optimal"] = n_optimal
        results["details"]["scenarios_failed"] = N_scenarios - n_optimal
        results["details"]["total_solve_time"] = round(sum(solve_times); digits=2)
        results["details"]["loop_wall_clock"] = round(loop_time; digits=2)
        results["details"]["mean_solve_time"] = round(sum(solve_times) / N_scenarios; digits=2)
        results["details"]["min_solve_time"] = round(minimum(solve_times); digits=2)
        results["details"]["max_solve_time"] = round(maximum(solve_times); digits=2)

        if !isempty(all_objectives)
            results["details"]["mean_objective"] = round(
                sum(all_objectives) / length(all_objectives); digits=2
            )
            results["details"]["min_objective"] = round(minimum(all_objectives); digits=2)
            results["details"]["max_objective"] = round(maximum(all_objectives); digits=2)
            results["details"]["objective_range"] = round(
                maximum(all_objectives) - minimum(all_objectives); digits=2
            )
        end

        results["details"]["scenario_1"] = scenario_results[1]
        results["details"]["scenario_20"] = scenario_results[N_scenarios]

        results["details"]["accepts_timeseries_programmatically"] = true
        results["details"]["scenario_loop_method"] = "deepcopy(data) + replicate() + modify loads/gens + solve_mn_opf()"
        results["details"]["per_scenario_overhead"] = "deepcopy + replicate (data dict operations, no file I/O)"
        results["details"]["results_collectable"] = true

        if n_optimal >= 16  # at least 80% should solve for SMALL (more likely to have infeasible)
            results["status"] = "pass"
        else
            push!(results["errors"], "Only $n_optimal of $N_scenarios scenarios solved optimally")
        end

    catch e
        push!(results["errors"], string(typeof(e), ": ", sprint(showerror, e)))
        push!(results["errors"], sprint(showerror, e, catch_backtrace()))
    finally
        results["wall_clock_seconds"] = round(time() - t0; digits=2)
    end

    return results
end

if abspath(PROGRAM_FILE) == @__FILE__
    result = run()
    println(JSON.json(result, 2))
end
